sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ListStateImpl.scala

```diff
@@ -49,101 +50,187 @@ class ListStateImpl[S](
-     val encodedKey = stateTypesEncoder.encodeGroupingKey()
-     val unsafeRowValuesIterator = store.valuesIterator(encodedKey, stateName)
-     new Iterator[S] {
-       override def hasNext: Boolean = {
-         unsafeRowValuesIterator.hasNext
-       }
-
-       override def next(): S = {
-         val valueUnsafeRow = unsafeRowValuesIterator.next()
-         stateTypesEncoder.decodeValue(valueUnsafeRow)
-       }
-     }
-   }
-
-   /** Update the value of the list. */
-   override def put(newState: Array[S]): Unit = {
-     validateNewState(newState)
-
-     val encodedKey = stateTypesEncoder.encodeGroupingKey()
-     var isFirst = true
-     var entryCount = 0L
-     TWSMetricsUtils.resetMetric(metrics, "numUpdatedStateRows")
-
-     newState.foreach { v =>
-       val encodedValue = stateTypesEncoder.encodeValue(v)
-       if (isFirst) {
-         store.put(encodedKey, encodedValue, stateName)
-         isFirst = false
-       } else {
-         store.merge(encodedKey, encodedValue, stateName)
-       }
-       entryCount += 1
-       TWSMetricsUtils.incrementMetric(metrics, "numUpdatedStateRows")
-     }
-     updateEntryCount(encodedKey, entryCount)
-   }
-
-   /** Append an entry to the list. */
-   override def appendValue(newState: S): Unit = {
-     StateStoreErrors.requireNonNullStateValue(newState, stateName)
-     val encodedKey = stateTypesEncoder.encodeGroupingKey()
-     val entryCount = getEntryCount(encodedKey)
-     store.merge(encodedKey,
-         stateTypesEncoder.encodeValue(newState), stateName)
-     TWSMetricsUtils.incrementMetric(metrics, "numUpdatedStateRows")
-     updateEntryCount(encodedKey, entryCount + 1)
-   }
-
-   /** Append an entire list to the existing value. */
-   override def appendList(newState: Array[S]): Unit = {
-     validateNewState(newState)
-
-     val encodedKey = stateTypesEncoder.encodeGroupingKey()
-     var entryCount = getEntryCount(encodedKey)
-     newState.foreach { v =>
-       val encodedValue = stateTypesEncoder.encodeValue(v)
-       store.merge(encodedKey, encodedValue, stateName)
-       entryCount += 1
-       TWSMetricsUtils.incrementMetric(metrics, "numUpdatedStateRows")
-     }
-     updateEntryCount(encodedKey, entryCount)
-   }
-
-   /** Remove this state. */
-   override def clear(): Unit = {
-     val encodedKey = stateTypesEncoder.encodeGroupingKey()
-     store.remove(encodedKey, stateName)
-     val entryCount = getEntryCount(encodedKey)
-     TWSMetricsUtils.incrementMetric(metrics, "numRemovedStateRows", entryCount)
-     removeEntryCount(encodedKey)
-   }
-
-   private def validateNewState(newState: Array[S]): Unit = {
-     StateStoreErrors.requireNonNullStateValue(newState, stateName)
-     StateStoreErrors.requireNonEmptyListStateValue(newState, stateName)
-
-     newState.foreach { v =>
-       StateStoreErrors.requireNonNullStateValue(v, stateName)
-     }
-   }
- }
+  override def exists(): Boolean = {
+    if (usingAvro) {
```

Comment:
I see a lot of duplicate code in all the functions in this file. Same with the ValueState implementation too. There should be a better way to do this without duplicating a lot of the code. The encoders both implement `StateTypesEncoder` and the only difference is one returns UnsafeRow and the other Array[Byte].

Also, the repeated `if (usingAvro)` check doesn't make the code clean. Can we look into if we can have a factory method or do something like this: 
`private val stateTypesEncoder: StateTypesEncoder  = if (usingAvro) new Avro else new Unsafe`

And then just use `stateTypesEncoder` in the methods instead of the repeated usingAvro check

Reply:
Yeah that's the first thing we've tried. The problem is if we do this, because `encodeGroupingKey` has different output types based on the encoder you used, Scala compilation failed when calling `store.get`